File: app/src/am_sonhk/com/example/asian/UserAdapter.java
package com.example.asian;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class UserAdapter extends RecyclerView.Adapter<UserAdapter.UserViewHolder> {

    private final List<User> mUserList;
    private final DatabaseHelper mDatabaseHelper;

    public UserAdapter(List<User> userList, DatabaseHelper databaseHelper) {
        this.mUserList = userList;
        this.mDatabaseHelper = databaseHelper;
    }

    @NonNull
    @Override
    public UserViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.user_item, parent, false);
        return new UserViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull UserViewHolder holder, int position) {
        User user = mUserList.get(position);
        holder.tvUserId.setText(String.valueOf(user.getUserId()));
        holder.tvUserName.setText(user.getUserName());
        holder.tvUserAge.setText(String.valueOf(user.getAge()));

        holder.btnDeleteUser.setOnClickListener(v -> {
            // Delete the user from the database
            mDatabaseHelper.deleteUserById(user.getUserId());
            // Remove the user from the list and notify RecyclerView by using DiffUtil
            List<User> newListUser = new ArrayList<>(mUserList);
            newListUser.remove(user);
            updateList(newListUser);
            // Show a toast message
            Toast.makeText(holder.itemView.getContext(), holder.itemView.getContext().getString(R.string.user_deleted), Toast.LENGTH_SHORT).show();
//            Toast.makeText()
        });
    }

    public void updateList(List<User> newList) {
        MyDiffUtilsCallback diffCallback = new MyDiffUtilsCallback(this.mUserList, newList);
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(diffCallback);
        mUserList.clear();
        mUserList.addAll(newList);
        diffResult.dispatchUpdatesTo(this);
    }


    @Override
    public int getItemCount() {
        return mUserList.size();
    }

    static class UserViewHolder extends RecyclerView.ViewHolder {
        TextView tvUserId;
        TextView tvUserName;
        TextView tvUserAge;
        Button btnDeleteUser;

        public UserViewHolder(@NonNull View itemView) {
            super(itemView);
            tvUserId = itemView.findViewById(R.id.tvUserId);
            tvUserName = itemView.findViewById(R.id.tvUserName);
            tvUserAge = itemView.findViewById(R.id.tvUserAge);
            btnDeleteUser = itemView.findViewById(R.id.btnDeleteUser);
        }
    }
}

File: app/src/am_sonhk/com/example/asian/kotlin/adapter/UserAdapter.kt
package com.example.asian.kotlin.adapter

import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.asian.databinding.UserItemBinding
import com.example.asian.kotlin.model.User

class UserAdapter(
    private val deleteUser: (User) -> Unit,
    private val updateUser: (User) -> Unit
) : ListAdapter<User, UserAdapter.UserViewHolder>(UserDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
        val binding = UserItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return UserViewHolder(binding)
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        val user = getItem(position)
        holder.bind(user, deleteUser, updateUser)

    }

    class UserViewHolder(private val binding: UserItemBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(user: User, deleteUser: (User) -> Unit, updateUser: (User) -> Unit) {
            binding.tvUserId.text = user.userId.toString()
            binding.tvUserName.text = user.userName
            binding.tvUserAge.text = user.userAge.toString()
            binding.btnDeleteUser.setOnClickListener {
                deleteUser(user)
            }
            binding.btnEditUser.setOnClickListener {
                updateUser(user.copy())
            }
        }
    }

    class UserDiffCallback : DiffUtil.ItemCallback<User>() {
        override fun areItemsTheSame(oldItem: User, newItem: User): Boolean {
            return oldItem.userId == newItem.userId
        }

        override fun areContentsTheSame(oldItem: User, newItem: User): Boolean {
            return oldItem == newItem
        }
    }
}

